File: packages/db/src/schema/approval-contract.ts
/**
 * Sprint C-12 (ADR 0042) — the shared approval/consent contract.
 *
 * One definition of "is this child's profile approved, by whom, under which
 * consent?", honoured by BOTH lesson pipelines:
 *
 *   - web-v2 (`apps/web-v2/lib/db/repos.ts: createLessonRun`) gates the live
 *     surface on it;
 *   - the services side (`services/learning-svc/.../sessions.ts` path-init,
 *     `services/brain-svc/.../routes/brain.py`) gates lesson/path init on it.
 *
 * There is no shared TS↔Python codegen in this repo, so the Python mirror
 * (`services/brain-svc/src/brain_svc/contracts/approval_contract.py`) restates
 * these exact constants and a parity test on each side fails CI on drift. ADR
 * 0042 is the human-readable parity record.
 *
 * This module is dependency-free (no drizzle/zod imports) so it can be consumed
 * anywhere — a route guard, a validator, a test — without dragging the schema
 * graph along. The canonical *table* shape lives in
 * `./brain-profile-approvals.ts` (migration 0108); this is its typed contract.
 */

/**
 * The full approval lifecycle of a learner's brain profile. Mirrors
 * `brainApprovalStatusEnum` (`./enums.ts`) and brain-svc's
 * `brain_states.approval_status`.
 */
export const BRAIN_APPROVAL_STATUSES = [
  "pending_parent_review",
  "approved",
  "amended",
  "declined",
] as const;
export type BrainApprovalStatus = (typeof BRAIN_APPROVAL_STATUSES)[number];

/**
 * The sub-set of statuses from which a lesson pipeline MAY teach. An amendment
 * is an accepted profile carrying parent overrides — it teaches. A profile that
 * is still `pending_parent_review` or has been `declined` MUST NOT teach.
 *
 * This is THE teach gate's decision set, asserted server-side against every
 * lesson pipeline (web `createLessonRun`, services path-init/lesson entry).
 */
export const TEACHING_APPROVAL_STATUSES = ["approved", "amended"] as const;
export type TeachingApprovalStatus = (typeof TEACHING_APPROVAL_STATUSES)[number];

/**
 * The action a single approval *record* (`brain_profile_approvals`) captures.
 * A decline is recorded too (it archives — C-06), but it is never a teaching
 * status. Distinct from the lifecycle status above: this is what the parent
 * *did*, recorded as audit evidence.
 */
export const APPROVAL_RECORD_ACTIONS = ["approved", "amended", "declined"] as const;
export type ApprovalRecordAction = (typeof APPROVAL_RECORD_ACTIONS)[number];

/**
 * The canonical field list of a shared approval record. Both stacks construct a
 * record with exactly these fields (web-v2 `BrainProfileApproval`; brain-svc
 * `BrainProfileApprovalRecord`). Kept as a literal so the parity tests can
 * assert the two stacks agree on the shape, not just the enums.
 */
export const APPROVAL_RECORD_FIELDS = [
  "id",
  "tenantId",
  "learnerId",
  "brainProfileId",
  "profileRevision",
  "actorUserId",
  "action",
  "consentVersion",
  "raiVersion",
  "modifications",
  "ipHash",
  "createdAt",
] as const;

/** The shared approval record contract (the C-06 `brain_profile_approvals` shape). */
export interface ApprovalContractRecord {
  id: string;
  tenantId: string;
  learnerId: string;
  brainProfileId: string;
  /** The `LearnerBrainProfile.revision` / brain_state version the parent reviewed. */
  profileRevision: number;
  actorUserId: string;
  action: ApprovalRecordAction;
  /** Version string of the COPPA consent copy the parent agreed to. */
  consentVersion: string;
  /** Version of the Responsible-AI disclosures acknowledged. */
  raiVersion: string;
  /** Field-level corrections folded at approval; [] for a plain approve / a decline. */
  modifications: unknown[];
  /** SHA-256 of the request IP — raw IP is never stored. May be null. */
  ipHash: string | null;
  createdAt: string;
}

/** True iff a profile in `status` is allowed to teach. THE teach-gate predicate. */
export function canTeach(status: string | null | undefined): status is TeachingApprovalStatus {
  return (
    status != null && (TEACHING_APPROVAL_STATUSES as readonly string[]).includes(status)
  );
}

/** True iff `status` is a recognised lifecycle status. */
export function isApprovalStatus(status: string | null | undefined): status is BrainApprovalStatus {
  return status != null && (BRAIN_APPROVAL_STATUSES as readonly string[]).includes(status);
}

/** True iff `action` is a recognised approval-record action. */
export function isApprovalRecordAction(
  action: string | null | undefined,
): action is ApprovalRecordAction {
  return action != null && (APPROVAL_RECORD_ACTIONS as readonly string[]).includes(action);
}

/**
 * Validate an arbitrary object against the shared approval-record contract.
 * Returns `{ ok: true, record }` on success, or `{ ok: false, errors }` listing
 * every violated field. Used by brain-svc's parity assertion and as a
 * defensive validator before persisting a record. Does not coerce — a wrong
 * type is an error, not a silent fix.
 */
export function parseApprovalRecord(
  input: unknown,
): { ok: true; record: ApprovalContractRecord } | { ok: false; errors: string[] } {
  const errors: string[] = [];
  if (typeof input !== "object" || input === null) {
    return { ok: false, errors: ["record must be a non-null object"] };
  }
  const o = input as Record<string, unknown>;
  const str = (k: string) => {
    if (typeof o[k] !== "string" || (o[k] as string).length === 0) {
      errors.push(`${k} must be a non-empty string`);
    }
  };
  str("id");
  str("tenantId");
  str("learnerId");
  str("brainProfileId");
  str("actorUserId");
  str("consentVersion");
  str("raiVersion");
  str("createdAt");
  if (typeof o.profileRevision !== "number" || !Number.isFinite(o.profileRevision)) {
    errors.push("profileRevision must be a finite number");
  }
  if (!isApprovalRecordAction(o.action as string)) {
    errors.push(`action must be one of ${APPROVAL_RECORD_ACTIONS.join(", ")}`);
  }
  if (!Array.isArray(o.modifications)) {
    errors.push("modifications must be an array");
  }
  if (o.ipHash !== null && typeof o.ipHash !== "string") {
    errors.push("ipHash must be a string or null");
  }
  if (errors.length > 0) return { ok: false, errors };
  return { ok: true, record: o as unknown as ApprovalContractRecord };
}

File: services/brain-svc/src/brain_svc/contracts/approval_contract.py
"""Sprint C-12 (ADR 0042) — the shared approval/consent contract, Python side.

The Python mirror of ``packages/db/src/schema/approval-contract.ts``. One
definition of "is this child's profile approved, by whom, under which consent?",
honoured by BOTH lesson pipelines. There is no shared TS<->Python codegen in this
repo, so these constants restate the TypeScript ones verbatim and a parity test
on each side fails CI on drift:

  - TS parity:  apps/web-v2/lib/db/__tests__/approval-contract.parity.test.ts
  - PY parity:  services/brain-svc/tests/test_approval_contract_parity.py

ADR 0042 is the human-readable parity record. Keep the literal sets below
identical to the TypeScript module.
"""
from __future__ import annotations

from typing import Any

from fastapi import HTTPException
from pydantic import BaseModel, field_validator

# ── Status enums (mirror of approval-contract.ts) ──────────────────────────

#: The full approval lifecycle of a learner's brain profile. Mirrors
#: ``brainApprovalStatusEnum`` and ``brain_states.approval_status``.
BRAIN_APPROVAL_STATUSES: tuple[str, ...] = (
    "pending_parent_review",
    "approved",
    "amended",
    "declined",
)

#: The sub-set of statuses from which a lesson pipeline MAY teach. An amendment
#: is an accepted profile carrying parent overrides — it teaches. This is THE
#: teach gate's decision set, asserted server-side against every lesson pipeline.
TEACHING_APPROVAL_STATUSES: tuple[str, ...] = ("approved", "amended")

#: The action a single approval *record* captures. A decline is recorded too
#: (it archives), but it is never a teaching status.
APPROVAL_RECORD_ACTIONS: tuple[str, ...] = ("approved", "amended", "declined")

#: The canonical field list of a shared approval record. Both stacks construct a
#: record with exactly these fields (camelCase — the wire/record shape).
APPROVAL_RECORD_FIELDS: tuple[str, ...] = (
    "id",
    "tenantId",
    "learnerId",
    "brainProfileId",
    "profileRevision",
    "actorUserId",
    "action",
    "consentVersion",
    "raiVersion",
    "modifications",
    "ipHash",
    "createdAt",
)


def can_teach(status: str | None) -> bool:
    """True iff a profile in ``status`` is allowed to teach. THE teach-gate predicate."""
    return status in TEACHING_APPROVAL_STATUSES


def is_approval_status(status: str | None) -> bool:
    """True iff ``status`` is a recognised lifecycle status."""
    return status in BRAIN_APPROVAL_STATUSES


def is_approval_record_action(action: str | None) -> bool:
    """True iff ``action`` is a recognised approval-record action."""
    return action in APPROVAL_RECORD_ACTIONS


def assert_can_teach(status: str | None) -> None:
    """Raise HTTP 403 unless ``status`` may teach.

    The explicit, reusable services-side gate (ADR 0042 §3). Any path into
    lesson/path init — internal-token callers included — runs this so the Python
    gate stops being sequencing-only. Detail string carries a stable
    ``brain_not_approved`` code, in parity with web-v2's typed
    ``createLessonRun`` result.
    """
    if not can_teach(status):
        raise HTTPException(
            status_code=403,
            detail={
                "code": "brain_not_approved",
                "message": (
                    "Brain profile awaiting parent approval before lesson "
                    "generation"
                ),
                "approvalStatus": status,
            },
        )


class BrainProfileApprovalRecord(BaseModel):
    """The shared approval record contract (the C-06 ``brain_profile_approvals``
    shape). Mirrors web-v2's ``BrainProfileApproval``. Field names are camelCase
    to match the record/wire shape exactly across both stacks.
    """

    id: str
    tenantId: str
    learnerId: str
    brainProfileId: str
    #: The ``LearnerBrainProfile.revision`` / brain_state version reviewed.
    profileRevision: int
    actorUserId: str
    action: str
    #: Version string of the COPPA consent copy the parent agreed to.
    consentVersion: str
    #: Version of the Responsible-AI disclosures acknowledged.
    raiVersion: str
    #: Field-level corrections folded at approval; [] for a plain approve / decline.
    modifications: list[Any] = []
    #: SHA-256 of the request IP — raw IP is never stored. May be None.
    ipHash: str | None = None
    createdAt: str

    @field_validator("action")
    @classmethod
    def _action_in_enum(cls, v: str) -> str:
        if not is_approval_record_action(v):
            raise ValueError(
                f"action must be one of {', '.join(APPROVAL_RECORD_ACTIONS)}"
            )
        return v
